sklearn/ensemble/_weight_boosting.py

```diff
@@ -1165,10 +1174,39 @@ def _boost(self, iboost, X, y, sample_weight, random_state):
         # Boost weight using AdaBoost.R2 alg
         estimator_weight = self.learning_rate * np.log(1.0 / beta)
 
-        if not iboost == self.n_estimators - 1:
+        if iboost == 0:
+            update_weight_method = "update"
+        elif iboost == self.n_estimators - 1:
+            update_weight_method = "do_nothing"
+        else:
+            if self.estimator_errors_[iboost - 1] > estimator_error:
+                if self.no_improvement == "reset_weights":
+                    update_weight_method = "reset"
+                elif self.no_improvement == "stop":
+                    warnings.warn(
+                        "Terminates training because the estimator error has increased."
+                    )
+                    return None, None, None
+                elif self.no_improvement == "warn":
+                    update_weight_method = "update"
+                    warnings.warn(
+                        "The estimator error has increased. "
+                        "Please consider hyper-parameter tuning for "
+                        "the base estimators."
+                    )
```

Comment:
I think this should raise `sklearn.exceptions.ConvergenceWarning` more specifically.